AutoGrade/views.py

```diff
@@ -269,46 +265,97 @@ def api(request, action):
                         response_data = {"status": 400, "type": "ERROR",
                          "message": "Assignment submission date expired"}
                     else:
-                        submission = Submission(submission_file=request.FILES['submission_file'],
-                            assignment=assignment,
-                            student=student)
-                        submission.save()
-
-                        submission_file_url = submission.submission_file.url
-                        extract_directory = submission_file_url.replace(".zip","/")
-
-                        zip_file = zipfile.ZipFile(submission.submission_file.url, 'r')
-                        zip_file.extractall(extract_directory)
-                        zip_file.close()
-
-                        # Move Instructor Test File
-                        shutil.copy(assignment.instructor_test.url, extract_directory)
-
-                        # Move Student Test File
-                        shutil.copy(assignment.student_test.url, extract_directory)
-        
-                        score, outlog = run_student_tests(extract_directory, assignment.total_points, assignment.timeout)
-                        
-                        submission.passed  = score[0]
-                        submission.failed  = score[1]
-                        submission.percent = score[2]
-
-                        submission.save()
-
-                        response_data = {"status": 200, "type": "SUCCESS",
-                             "message": score}
-
-                else:
+                        # Solve bruteforcing...
```

Comment:
Hnn, let's not merge this for now. I have some questions. (a) this looks like the queue is processed at the call to the API. What happens if immediately after a submission, the connection is broken. Do we have a nohup type scenario so that the queue remains processing? 

If things are left in the queue, we will probably need a cron/daemon to process the queue, no? 

Also, how is the student going to get "real time" feedback if we are going to be doing async queue processing? 

And we will have to do some rigorous resting before this is merged. Nice work on this but I have had some headaches with such queues and I don't want to have to handle them on prod =]

Reply:
Every time student submits the assignment, their submission will be processed on the spot. However if number of current processing is equals to available cores then submission will set to status of QUEUE and student will be showed this message "Your submission is added to queue, you can view result in web interface." (Student will not be able to submit any submission until his/her submission is in QUEUE or EXECUTING state).

Another column is added to Course (Submission History) section which indicates the status of their submission.

After the successful processing of submission result is returned, plus a Thread is started for handling QUEUE submissions. If there is some item in QUEUE then there will a EXECUTING process running also which will handle the QUEUE item later with Thread.

Yes, there should be done some rigorous testing before moving this to prod.

Reply:
Well, if there's something in the queue and connection is broken, does it keep processing? Doesn't happen in the normal cases unless you took care of it somehow.

Reply:
In case Django server is stopped then this case need to be handle :) Will look into this later.

Reply:
Don't think it even needs the whole server to stop. As far as my understanding goes, if the client is disconnected, all threads for that are stopped. (Anyway, let's worry about other stuff first.)